tests/includes/juju.sh

```diff
@@ -218,30 +217,53 @@ juju_bootstrap() {
 		;;
 	esac
 
-	image_metadata=""
+	pre_bootstrap
+
+	# When double quotes are added to ${series}, the juju bootstrap
+	# command looks correct, and works outside of the harness, but
+	# does not run, goes directly to cleanup.
+	command="juju bootstrap ${series} --build-agent=${BUILD_AGENT} ${cloud} ${name} -d ${model} ${BOOTSTRAP_ADD_ON}"
+	# keep $@ here, otherwise hit SC2124
+	${command} "$@" 2>&1 | OUTPUT "${output}"
+	echo "${name}" >>"${TEST_DIR}/jujus"
+
+	post_bootstrap
+}
 
-	if [[ ${BOOTSTRAP_PROVIDER} == "vsphere" ]]; then
+# pre_bootstrap contains setup required before bootstrap specific to providers
+# and shouldn't be used by any of the tests directly.
+pre_bootstrap() {
+	case "${BOOTSTRAP_PROVIDER:-}" in
+	"vsphere")
 		echo "====> Creating image simplestream metadata for juju ($(green "${version}:${cloud}"))"
 
 		image_streams_dir=${TEST_DIR}/image-streams
 		setup_vsphere_simplestreams "${image_streams_dir}" "${BOOTSTRAP_SERIES}"
-		image_metadata="--metadata-source ${image_streams_dir}"
-	fi
-
-	# When double quotes are added to ${series}, the juju bootstrap
-	# command looks correct, and works outside of the harness, but
-	# does not run, goes directly to cleanup.
-	#shellcheck disable=SC2086
-	juju bootstrap ${series} \
-		--build-agent=${BUILD_AGENT} \
-		"${cloud}" "${name}" -d "${model}" ${image_metadata} "$@" 2>&1 | OUTPUT "${output}"
+		export BOOTSTRAP_ADD_ON="--metadata-source ${image_streams_dir}"
+		;;
+	esac
+}
 
-	echo "${name}" >>"${TEST_DIR}/jujus"
+# post_bootstrap contains actions required after bootstrap specific to providers
+# and shouldn't be used by any of the tests directly.  Calls post_add_model
+# models are added during bootstrap.
+post_bootstrap() {
```

Comment:
We might want to put post_bootstrap into a bash 'trap' so that it always runs whether or not bootstrap succeeds. Then again, maybe that happens too late in the happy case?

Reply:
At least in this case, it's easier to debug bootstrap failing later if post_bootstrap does not run.  Something to consider if in the future if options are added.